runtime/src/address_map/cache.rs

```diff
@@ -0,0 +1,581 @@
+                    .map(lookup_address)
+                    .collect::<Option<Vec<_>>>()?,
+            );
+        }
+
+        Some(mapped_addresses)
+    }
+
+    /// Insert newly activated map into the cache
+    fn insert_activated_map(
+        &self,
+        key: Pubkey,
+        entries: Arc<Vec<Pubkey>>,
+        activation_epoch: Epoch,
+    ) {
+        let cached_address_map = CachedAddressMap {
+            entries,
+            activation_epoch,
+            deactivation_epoch: None,
+        };
+
+        if self.address_maps.insert(key, cached_address_map).is_some() {
+            warn!("Overwrote activated address map: {}", key);
+        }
+    }
+
+    /// Update the address map cache with the pending changes from a rooted
+    /// bank. Use the bank's epoch to record when address maps have started
+    /// warming up or cooling down for new activations and deactivations,
+    /// respectively. Also use the epoch for pruning any deactivated maps that
+    /// have finished cooling down.
+    pub fn update_cache(
+        &self,
+        rooted_bank_changes: &AddressMapPendingChanges,
+        rooted_bank_epoch: Epoch,
+    ) {
+        rooted_bank_changes
+            .activations_iter()
+            .for_each(|activated_map| {
+                self.insert_activated_map(
+                    *activated_map.key(),
+                    Arc::clone(activated_map.value()),
+                    rooted_bank_epoch,
+                );
+            });
+
+        rooted_bank_changes
+            .deactivations_iter()
+            .for_each(|deactivated_map| {
+                if let Some(mut address_map) = self.address_maps.get_mut(deactivated_map.key()) {
+                    address_map.value_mut().deactivation_epoch = Some(rooted_bank_epoch);
+                } else {
+                    warn!(
+                        "Failed to find address map: {} to deactivate",
+                        deactivated_map.key()
+                    );
+                }
+            });
+
+        // Only prune when we encounter a new epoch to avoid doing unnecessary work for each bank
+        if self
+            .last_rooted_bank_epoch
+            .swap(rooted_bank_epoch, Ordering::Relaxed)
+            < rooted_bank_epoch
+        {
+            self.prune_deactivated_maps(rooted_bank_epoch);
+        }
+    }
+
+    /// Prune all cached address maps that have been deactivated and passed the cooldown period
+    fn prune_deactivated_maps(&self, rooted_bank_epoch: Epoch) {
+        self.address_maps.retain(|_address, address_map| {
+            if let Some(deactivation_epoch) = address_map.deactivation_epoch {
+                rooted_bank_epoch < deactivation_epoch.saturating_add(DEACTIVATION_COOLDOWN)
+            } else {
+                true
+            }
+        });
+    }
+
+    /// Called during snapshot processing to populate the cache from rooted slots
+    pub fn populate_cache(&self, map_key: &Pubkey, map_data: &[u8]) {
```

Comment:
Any reason not to make this fallible?  Seems like it'd help debugging deserialization failures

Reply:
No reason, updated to return a result